framework/src/interface/packet.rs

```diff
@@ -436,12 +436,16 @@ impl<T: EndOffset, M: Sized + Send> Packet<T, M> {
     pub fn insert_header_fn<T2: EndOffset<PreviousHeader = T>>(
         &mut self,
         header: &T2,
-        on_insert: &Fn(&mut T, isize),
-    ) -> Result<isize> {
+        on_insert: &Fn(&mut T),
+    ) -> Result<isize>
+    where
+        T: HeaderUpdates,
+    {
         unsafe {
             if let Ok(diff) = self.insert_header(header) {
                 let current_header: &mut T = &mut *self.header();
-                on_insert(current_header, diff);
+                current_header.update_payload(diff);
+                on_insert(current_header);
```

Comment:
The other thing that's very common (universal?) when inserting/removing headers is updating the `next_header` field. May want to think about doing that automatically within each of these. It's another scenario where, unless the caller remembers to do it, would leave the packet invalid.

Reply:
Yep. Will try to get these semantics right.